Wox.Infrastructure/Image/WindowsThumbnailCacheAPI.cs

```diff
@@ -0,0 +1,227 @@
+﻿
+using System;
+using System.ComponentModel;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Windows;
+using System.Windows.Interop;
+using System.Windows.Media;
+using System.Windows.Media.Imaging;
+
+namespace Wox.Infrastructure.Image 
+{
+    static class WindowsThumbnailCacheAPI 
+    {
+        private static readonly IThumbnailCache _cache;
+
+        public static readonly HashSet<string> NotCachableExtensions;
+
+        static WindowsThumbnailCacheAPI() {
+            NotCachableExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase) {
+                ".lnk",
+                ".exe",
+                ".bat",
+                ".cmd",
+                ".pif",
+                ".appref-ms"
+            };
+
+            Guid CLSID_LocalThumbnailCache = new Guid("50EF4544-AC9F-4A8E-B21B-8A26180DB13F");
+            _cache = (IThumbnailCache)Activator.CreateInstance(Type.GetTypeFromCLSID(CLSID_LocalThumbnailCache));
+        }
+
+        public static ImageSource GetThumbnail(string path, int size) {
+            IShellItem item = GetShellItemFromPath(path);
+            if (item == null) {
+                return null;
+            }
+
+
+            ISharedBitmap bitmap = null;
+            IntPtr hBitmap = IntPtr.Zero;
+            try {
+
+                
+                WTS_CACHEFLAGS cacheFlags;
+                WTS_THUMBNAILID thumbid;
+
+                uint ret;
+
+                ret = _cache.GetThumbnail(item, (uint)size, WTS_FLAGS.WTS_EXTRACT, out bitmap, out cacheFlags, out thumbid);
+
+                if (ret != 0 || bitmap == null) {
+                    return null;
```

Comment:
So if it matches the extension pattern but the thumbnail does not return an image, we just leave it blank?